src/ray/common/asio/instrumented_io_context.h

```diff
@@ -28,7 +28,31 @@ class instrumented_io_context : public boost::asio::io_context {
  public:
   /// Initializes the global stats struct after calling the base contructor.
   /// TODO(ekl) allow taking an externally defined event tracker.
-  instrumented_io_context() : event_stats_(std::make_shared<EventTracker>()) {}
+  instrumented_io_context()
+      : is_running_{false}, event_stats_(std::make_shared<EventTracker>()) {}
+
+  /// Run the io_context if and only if no other thread is running it. Blocks
+  /// other threads from running once started. Noop if there is a thread
+  /// running it already.
+  void run_exclusive(std::function<void()> callback) {
+    absl::MutexLock l(&mu_);
+    // Note: this doesn't set is_running_ because it blocks anything else from running
+    // anyway.
+    if (!is_running_) {
+      callback();
+      boost::asio::io_context::run();
+    }
+  }
+
+  void run() {
```

Comment:
honestly if the callbacks were guaranteed thread-safe, I would prefer the approach from @iycheng 

but yeah, ideally this can plug the gap until gcs has executors